docs/dev/migrating-to-v1.rst

```diff
@@ -235,6 +239,19 @@ In the core of HydroMT, the available components are (list or maybe table):
 
  A user can defined its own new component either by inheriting from the base ``ModelComponent`` or from another one (eg SubgridComponent(GridComponent)). The new components can be accessed and discovered through the `PLUGINS` architecture of HydroMT similar to Model plugins. See the related paragraph for more details.
 
+The `Model.__init__` function can be used to add default components by plugins like so:
+
+```python
+
+class ExampleModel(Model):
+	def __init__(self):
+        self.root: ModelRoot = ModelRoot(".")
+		self.add_component("region", ModelRegionComponent)
+		self.add_component("grid", GridComponent)
+		...
+
+```
+
```

Comment:
```suggestion

If you want to allow your plugin user to modify the root and update or add new component during instantiation then you can use:

``` python

class ExampleEditModel(Model):
    def __init__(
        self,
        components: Optional[dict[str, dict[str, Any]]] = None,
        root: Optional[str] = None,
    ):        
        # Recursively update the components with any defaults that are missing in the components provided by the user.
        components = components or {}
        default_components = {
            "region": {"type": "ModelRegionComponent"},
            "grid": {"type": GridComponent},
        }
        components = hydromt._utils.deep_merge.deep_merge(
            default_components, components
        )
        
        # Now instantiate the Model
        super().__init__(
            root = root,
            components = components,
        )

```

While writing this example that plugins will actually use, this means ``deep_merge`` and ``_utils`` will not be fully private API of hydromt. Should we then maybe have a public ``utils`` and a private ``_utils`` ?
The other you can do this for plugins is by adding a default_components argument to ``Model.__init__`` so that plugins could then just do instead:

``` python

class ExampleEditModel(Model):
    def __init__(
        self,
        components: Optional[dict[str, dict[str, Any]]] = None,
        root: Optional[str] = None,
    ):
        default_components = {
            "region": {"type": "ModelRegionComponent"},
            "grid": {"type": GridComponent},
        }
        
        super().__init__(
            root = root,
            components = components,
            default_components = default_components,
        )

``` 
I think I still prefer the first solution which is more transparent in a way. @savente93 what do you think?

Reply:
yeah I agree, I made a public `utils` module and placed the `deep_merge` there. I'll see if I can quicly fix taking the default components as arguments.

Reply:
It took me a second to realize but the code as you wrote it should already work, and I do think it's a better example. so we should be all good now.